File: prepare_dataset.m
%% prepare_dataset.m
% Extract RAR files into class folders and prepare them for MATLAB training.
% Outputs:
%   1) extracted_data/      : extracted raw data
%   2) prepared_data/       : image copies grouped by label
%   3) dataset_for_training.mat : train/validation/test data

clear; clc;

%% Settings
projectDir = fileparts(mfilename("fullpath"));
archivePattern = "*.rar";

extractedDir = fullfile(projectDir, "extracted_data");
preparedDir = fullfile(projectDir, "prepared_data");
outputMat = fullfile(projectDir, "dataset_for_training.mat");

imageSize = [224 224 3];      % Change to [64 64 3], [128 128 3], etc. if needed.
trainRatio = 0.70;
valRatio = 0.15;
testRatio = 0.15;
customExtractorPath = "";     % Example: "C:\Program Files\7-Zip\7z.exe"

rng(42);                      % Reproducible split

%% 1. Find archives
archives = dir(fullfile(projectDir, archivePattern));
if isempty(archives)
    error("No %s files found in: %s", archivePattern, projectDir);
end

if ~exist(extractedDir, "dir")
    mkdir(extractedDir);
end
if ~exist(preparedDir, "dir")
    mkdir(preparedDir);
end

%% 2. Extract RAR files
extractor = findExtractor(customExtractorPath);
fprintf("Found %d archive files.\n", numel(archives));

for i = 1:numel(archives)
    archivePath = fullfile(archives(i).folder, archives(i).name);
    [~, className] = fileparts(archives(i).name);
    classExtractDir = fullfile(extractedDir, className);

    if exist(classExtractDir, "dir") && ~isempty(dir(fullfile(classExtractDir, "**", "*.*")))
        fprintf("[%s] already extracted. Skipping.\n", className);
        continue;
    end

    if extractor.kind == "none"
        fprintf("No RAR extractor found. Skipping automatic extraction.\n");
        break;
    end

    if ~exist(classExtractDir, "dir")
        mkdir(classExtractDir);
    end

    fprintf("[%s] extracting...\n", archives(i).name);
    extractArchive(archivePath, classExtractDir, extractor);
end

%% 3. Build file index
allFiles = listFiles(extractedDir);
if isempty(allFiles)
    error(["No extracted files found.\n" ...
           "Install 7-Zip/WinRAR, or set customExtractorPath near the top of this script,\n" ...
           "or manually extract each RAR into extracted_data/<class_name>/ and run again.\n" ...
           "Example folder: extracted_data/B/, extracted_data/FTW/"]);
end

labels = labelsFromTopFolder(allFiles, extractedDir);
fileTable = table(string(allFiles(:)), categorical(labels(:)), ...
    'VariableNames', ["File", "Label"]);

writetable(fileTable, fullfile(projectDir, "file_index.csv"));
fprintf("Saved file index: file_index.csv (%d files)\n", height(fileTable));

%% 4. If the data contains images, build image datastores
imageExts = [".jpg", ".jpeg", ".png", ".bmp", ".tif", ".tiff", ".gif"];
[~, ~, exts] = cellfun(@fileparts, allFiles, 'UniformOutput', false);
isImage = ismember(lower(string(exts)), imageExts);

if any(isImage)
    imageFiles = string(allFiles(isImage));
    imageLabels = labelsFromTopFolder(cellstr(imageFiles), extractedDir);

    makePreparedImageFolders(imageFiles, imageLabels, preparedDir);

    imds = imageDatastore(preparedDir, ...
        "IncludeSubfolders", true, ...
        "LabelSource", "foldernames", ...
        "ReadFcn", @(x)readAndFixImage(x, imageSize));

    [imdsTrain, imdsRest] = splitEachLabel(imds, trainRatio, "randomized");
    restValRatio = valRatio / (valRatio + testRatio);
    [imdsValidation, imdsTest] = splitEachLabel(imdsRest, restValRatio, "randomized");

    adsTrain = augmentedImageDatastore(imageSize(1:2), imdsTrain);
    adsValidation = augmentedImageDatastore(imageSize(1:2), imdsValidation);
    adsTest = augmentedImageDatastore(imageSize(1:2), imdsTest);

    save(outputMat, ...
        "imds", "imdsTrain", "imdsValidation", "imdsTest", ...
        "adsTrain", "adsValidation", "adsTest", ...
        "fileTable", "imageSize", "-v7.3");

    fprintf("\nImage training dataset created.\n");
    fprintf("Train: %d, Validation: %d, Test: %d\n", ...
        numel(imdsTrain.Files), numel(imdsValidation.Files), numel(imdsTest.Files));
    fprintf("Saved file: %s\n", outputMat);
    return;
end

%% 5. Otherwise, try to convert numeric files into a feature matrix
fprintf("No image files found. Trying to build a numeric feature matrix.\n");

[X, Y, usedFiles] = buildNumericFeatureMatrix(fileTable);
if isempty(X)
    error("Could not find image or numeric data files. Check file_index.csv.");
end

cv1 = cvpartition(Y, "HoldOut", 1 - trainRatio);
idxTrain = training(cv1);
idxRest = test(cv1);

YRest = Y(idxRest);
cv2 = cvpartition(YRest, "HoldOut", testRatio / (valRatio + testRatio));
restIndices = find(idxRest);
idxValidation = false(size(Y));
idxTest = false(size(Y));
idxValidation(restIndices(training(cv2))) = true;
idxTest(restIndices(test(cv2))) = true;

XTrain = X(idxTrain, :);
YTrain = Y(idxTrain);
XValidation = X(idxValidation, :);
YValidation = Y(idxValidation);
XTest = X(idxTest, :);
YTest = Y(idxTest);

save(outputMat, ...
    "X", "Y", "XTrain", "YTrain", "XValidation", "YValidation", "XTest", "YTest", ...
    "usedFiles", "fileTable", "-v7.3");

fprintf("\nNumeric training dataset created.\n");
fprintf("X size: %d x %d\n", size(X, 1), size(X, 2));
fprintf("Train: %d, Validation: %d, Test: %d\n", ...
    numel(YTrain), numel(YValidation), numel(YTest));
fprintf("Saved file: %s\n", outputMat);

%% Local functions
function extractor = findExtractor(customExtractorPath)
    if strlength(customExtractorPath) > 0 && isfile(customExtractorPath)
        [~, name] = fileparts(customExtractorPath);
        if contains(lower(name), "winrar")
            kind = "winrar";
        elseif contains(lower(name), "unrar")
            kind = "unrar";
        else
            kind = "7z";
        end
        extractor = struct("kind", kind, "path", customExtractorPath);
        return;
    end

    candidates = [
        "7z", "7z";
        "7z", "C:\Program Files\7-Zip\7z.exe";
        "7z", "C:\Program Files\NVIDIA Corporation\NVIDIA GeForce Experience\7z.exe";
        "winrar", "C:\Program Files\WinRAR\WinRAR.exe";
        "winrar", "C:\Program Files (x86)\WinRAR\WinRAR.exe";
        "unrar", "unrar"
    ];

    extractor = struct("kind", "none", "path", "");
    for k = 1:size(candidates, 1)
        kind = candidates(k, 1);
        path = candidates(k, 2);

        if contains(path, filesep) || contains(path, ":")
            if isfile(path)
                extractor.kind = kind;
                extractor.path = path;
                return;
            end
        else
            [status, out] = system("where " + path);
            if status == 0
                lines = splitlines(strtrim(out));
                extractor.kind = kind;
                extractor.path = string(lines(1));
                return;
            end
        end
    end
end

function extractArchive(archivePath, outDir, extractor)
    switch extractor.kind
        case "7z"
            cmd = sprintf('"%s" x -y -o"%s" "%s"', extractor.path, outDir, archivePath);
        case "winrar"
            cmd = sprintf('"%s" x -ibck -y "%s" "%s\\"', extractor.path, archivePath, outDir);
        case "unrar"
            cmd = sprintf('"%s" x -y "%s" "%s\\"', extractor.path, archivePath, outDir);
        otherwise
            error("Unsupported extractor.");
    end

    [status, out] = system(cmd);
    if status ~= 0
        error("Extraction failed:\n%s\n%s", archivePath, out);
    end
end

function files = listFiles(rootDir)
    d = dir(fullfile(rootDir, "**", "*"));
    d = d(~[d.isdir]);
    files = fullfile({d.folder}, {d.name});
    files = files(:);
end

function labels = labelsFromTopFolder(files, rootDir)
    labels = strings(numel(files), 1);
    rootDir = char(rootDir);

    for k = 1:numel(files)
        rel = erase(string(files{k}), string(rootDir) + filesep);
        parts = split(rel, filesep);
        labels(k) = parts(1);
    end
end

function makePreparedImageFolders(imageFiles, imageLabels, preparedDir)
    if ~exist(preparedDir, "dir")
        mkdir(preparedDir);
    end

    for k = 1:numel(imageFiles)
        label = string(imageLabels(k));
        labelDir = fullfile(preparedDir, label);
        if ~exist(labelDir, "dir")
            mkdir(labelDir);
        end

        [~, name, ext] = fileparts(imageFiles(k));
        dst = fullfile(labelDir, name + "_" + k + ext);
        if ~isfile(dst)
            copyfile(imageFiles(k), dst);
        end
    end
end

function img = readAndFixImage(filename, imageSize)
    img = imread(filename);

    if size(img, 3) == 1 && imageSize(3) == 3
        img = repmat(img, 1, 1, 3);
    elseif size(img, 3) == 4
        img = img(:, :, 1:3);
    end

    img = imresize(img, imageSize(1:2));
end

function [X, Y, usedFiles] = buildNumericFeatureMatrix(fileTable)
    vectors = {};
    labels = strings(0, 1);
    usedFiles = strings(0, 1);

    for k = 1:height(fileTable)
        file = fileTable.File(k);
        label = string(fileTable.Label(k));

        try
            v = readNumericVector(file);
        catch
            continue;
        end

        if isempty(v) || any(~isfinite(v))
            continue;
        end

        vectors{end + 1, 1} = v(:)'; %#ok<AGROW>
        labels(end + 1, 1) = label; %#ok<AGROW>
        usedFiles(end + 1, 1) = file; %#ok<AGROW>
    end

    if isempty(vectors)
        X = [];
        Y = categorical.empty(0, 1);
        return;
    end

    maxLen = max(cellfun(@numel, vectors));
    X = zeros(numel(vectors), maxLen);
    for k = 1:numel(vectors)
        v = vectors{k};
        X(k, 1:numel(v)) = v;
    end

    Y = categorical(labels);
    X = normalize(X, 2);
end

function v = readNumericVector(file)
    [~, ~, ext] = fileparts(file);
    ext = lower(ext);

    switch ext
        case {".csv", ".txt", ".dat", ".tsv"}
            opts = detectImportOptions(file);
            tbl = readtable(file, opts);
            numericVars = varfun(@isnumeric, tbl, "OutputFormat", "uniform");
            data = table2array(tbl(:, numericVars));
            v = data(:);
        case ".mat"
            s = load(file);
            names = fieldnames(s);
            v = [];
            for i = 1:numel(names)
                value = s.(names{i});
                if isnumeric(value)
                    v = [v; numericToFeatureVector(value)]; %#ok<AGROW>
                end
            end
        case ".wav"
            [audio, fs] = audioread(file);
            audio = mean(audio, 2);
            coeff = mfcc(audio, fs);
            v = coeff(:);
        otherwise
            v = [];
    end
end

function v = numericToFeatureVector(x)
    x = double(x);

    if isvector(x)
        x = x(:);
    end

    % Small real arrays can be used directly. Large time-domain signals are
    % summarized so the final X matrix stays trainable on a normal PC.
    if isreal(x) && numel(x) <= 20000
        v = x(:);
        return;
    end

    if isvector(x)
        x = x(:).';
    end

    mag = abs(x);
    xr = real(x);
    xi = imag(x);

    magMean = mean(mag, 2);
    magStd = std(mag, 0, 2);
    magMax = max(mag, [], 2);
    magRms = sqrt(mean(mag.^2, 2));
    realMean = mean(xr, 2);
    realStd = std(xr, 0, 2);
    imagMean = mean(xi, 2);
    imagStd = std(xi, 0, 2);

    v = [magMean; magStd; magMax; magRms; realMean; realStd; imagMean; imagStd];
end

File: prepare_dataset.py
#%%
from __future__ import annotations

import csv
import json
import shutil
import subprocess
from pathlib import Path

import numpy as np
from scipy.io import loadmat


PROJECT_DIR = Path(__file__).resolve().parent
EXTRACTED_DIR = PROJECT_DIR / "extracted_data"
OUT_DIR = PROJECT_DIR / "python_dataset"
TRAIN_RATIO = 0.70
VAL_RATIO = 0.15
TEST_RATIO = 0.15
RANDOM_SEED = 42

# If auto-detection fails, set this manually.
CUSTOM_EXTRACTOR = ""

# After python_dataset/*.npy and dataset.npz are saved, remove extracted .mat
# files under extracted_data/. Original .rar files are never deleted.
DELETE_EXTRACTED_MAT_AFTER_CONVERSION = True


def main() -> None:
    archives = sorted(PROJECT_DIR.glob("*.rar"))
    if not archives:
        raise FileNotFoundError(f"No .rar files found in {PROJECT_DIR}")

    EXTRACTED_DIR.mkdir(exist_ok=True)
    OUT_DIR.mkdir(exist_ok=True)

    extractor = find_extractor()
    extract_archives(archives, extractor)

    rows = collect_mat_files(archives)
    if not rows:
        raise RuntimeError(
            "No .mat files found. Check extraction or manually extract RAR files into extracted_data/<label>/."
        )

    label_names = sorted({row["label"] for row in rows})
    label_to_id = {name: idx for idx, name in enumerate(label_names)}

    X_list: list[np.ndarray] = []
    y_list: list[int] = []
    metadata: list[dict[str, str | int]] = []

    print(f"Found {len(rows)} .mat files across {len(label_names)} classes.")
    for idx, row in enumerate(rows, start=1):
        mat_path = Path(row["path"])
        label = str(row["label"])
        feature = mat_to_feature_vector(mat_path)

        X_list.append(feature.astype(np.float32))
        y_list.append(label_to_id[label])
        metadata.append(
            {
                "index": idx - 1,
                "file": str(mat_path),
                "label": label,
                "label_id": label_to_id[label],
            }
        )

        if idx % 25 == 0 or idx == len(rows):
            print(f"Converted {idx}/{len(rows)} files")

    X = np.vstack(X_list).astype(np.float32)
    y = np.asarray(y_list, dtype=np.int64)

    train_idx, val_idx, test_idx = stratified_split(y)

    save_array("X.npy", X)
    save_array("y.npy", y)
    save_array("X_train.npy", X[train_idx])
    save_array("y_train.npy", y[train_idx])
    save_array("X_val.npy", X[val_idx])
    save_array("y_val.npy", y[val_idx])
    save_array("X_test.npy", X[test_idx])
    save_array("y_test.npy", y[test_idx])

    np.savez_compressed(
        OUT_DIR / "dataset.npz",
        X=X,
        y=y,
        X_train=X[train_idx],
        y_train=y[train_idx],
        X_val=X[val_idx],
        y_val=y[val_idx],
        X_test=X[test_idx],
        y_test=y[test_idx],
        label_names=np.asarray(label_names),
    )

    write_json(OUT_DIR / "label_map.json", label_to_id)
    write_metadata(OUT_DIR / "metadata.csv", metadata)

    if DELETE_EXTRACTED_MAT_AFTER_CONVERSION:
        deleted_count = delete_extracted_mat_files()
        print(f"Deleted extracted .mat files: {deleted_count}")

    print("")
    print("Python dataset created:")
    print(f"  {OUT_DIR / 'dataset.npz'}")
    print(f"  X shape: {X.shape}")
    print(f"  train/val/test: {len(train_idx)}/{len(val_idx)}/{len(test_idx)}")
    print(f"  labels: {label_to_id}")


def find_extractor() -> Path | None:
    candidates = [
        CUSTOM_EXTRACTOR,
        shutil.which("7z"),
        shutil.which("WinRAR"),
        shutil.which("unrar"),
        r"C:\Program Files\7-Zip\7z.exe",
        r"C:\Program Files\NVIDIA Corporation\NVIDIA GeForce Experience\7z.exe",
        r"C:\Program Files\WinRAR\WinRAR.exe",
        r"C:\Program Files (x86)\WinRAR\WinRAR.exe",
    ]

    for item in candidates:
        if item and Path(item).is_file():
            return Path(item)
    return None


def extract_archives(archives: list[Path], extractor: Path | None) -> None:
    if extractor is None:
        print("No extractor found. Using already-extracted files only.")
        return

    print(f"Using extractor: {extractor}")
    for archive in archives:
        label = archive.stem
        target_dir = EXTRACTED_DIR / label
        if target_dir.exists() and any(target_dir.rglob("*.mat")):
            print(f"[{label}] already extracted")
            continue

        target_dir.mkdir(parents=True, exist_ok=True)
        print(f"[{archive.name}] extracting...")

        name = extractor.name.lower()
        if "winrar" in name:
            cmd = [str(extractor), "x", "-ibck", "-y", str(archive), str(target_dir) + "\\"]
        elif "unrar" in name:
            cmd = [str(extractor), "x", "-y", str(archive), str(target_dir) + "\\"]
        else:
            cmd = [str(extractor), "x", "-y", f"-o{target_dir}", str(archive)]

        subprocess.run(cmd, check=True)


def collect_mat_files(archives: list[Path]) -> list[dict[str, str]]:
    rows: list[dict[str, str]] = []
    for archive in archives:
        label = archive.stem
        class_dir = EXTRACTED_DIR / label
        if not class_dir.exists():
            continue

        for mat_path in sorted(class_dir.rglob("*.mat")):
            rows.append({"path": str(mat_path), "label": label})
    return rows


def mat_to_feature_vector(mat_path: Path) -> np.ndarray:
    data = loadmat(mat_path)
    arrays = [
        value
        for key, value in data.items()
        if not key.startswith("__") and isinstance(value, np.ndarray) and np.issubdtype(value.dtype, np.number)
    ]
    if not arrays:
        raise ValueError(f"No numeric arrays in {mat_path}")

    features = [array_to_features(arr) for arr in arrays]
    return np.concatenate(features)


def array_to_features(arr: np.ndarray) -> np.ndarray:
    x = np.asarray(arr)
    if x.ndim == 1:
        x = x.reshape(1, -1)
    elif x.ndim > 2:
        x = x.reshape(x.shape[0], -1)

    mag = np.abs(x).astype(np.float64)
    real = np.real(x).astype(np.float64)
    imag = np.imag(x).astype(np.float64)

    feature_blocks = [
        mag.mean(axis=1),
        mag.std(axis=1),
        mag.max(axis=1),
        np.sqrt(np.mean(mag**2, axis=1)),
        real.mean(axis=1),
        real.std(axis=1),
        imag.mean(axis=1),
        imag.std(axis=1),
    ]

    return np.concatenate(feature_blocks)


def stratified_split(y: np.ndarray) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    rng = np.random.default_rng(RANDOM_SEED)
    train_parts = []
    val_parts = []
    test_parts = []

    for label_id in np.unique(y):
        idx = np.flatnonzero(y == label_id)
        rng.shuffle(idx)

        n = len(idx)
        n_train = int(round(n * TRAIN_RATIO))
        n_val = int(round(n * VAL_RATIO))

        train_parts.append(idx[:n_train])
        val_parts.append(idx[n_train : n_train + n_val])
        test_parts.append(idx[n_train + n_val :])

    train_idx = np.concatenate(train_parts)
    val_idx = np.concatenate(val_parts)
    test_idx = np.concatenate(test_parts)

    rng.shuffle(train_idx)
    rng.shuffle(val_idx)
    rng.shuffle(test_idx)

    return train_idx, val_idx, test_idx


def save_array(name: str, array: np.ndarray) -> None:
    path = OUT_DIR / name
    np.save(path, array)
    print(f"Saved {path.name}: {array.shape}")


def write_json(path: Path, data: dict[str, int]) -> None:
    path.write_text(json.dumps(data, indent=2), encoding="utf-8")


def write_metadata(path: Path, rows: list[dict[str, str | int]]) -> None:
    with path.open("w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=["index", "file", "label", "label_id"])
        writer.writeheader()
        writer.writerows(rows)


def delete_extracted_mat_files() -> int:
    deleted = 0
    root = EXTRACTED_DIR.resolve()

    for mat_path in sorted(EXTRACTED_DIR.rglob("*.mat")):
        resolved = mat_path.resolve()
        if root not in resolved.parents:
            raise RuntimeError(f"Refusing to delete outside extracted_data: {resolved}")
        resolved.unlink()
        deleted += 1

    for directory in sorted(EXTRACTED_DIR.rglob("*"), reverse=True):
        if directory.is_dir():
            try:
                directory.rmdir()
            except OSError:
                pass

    return deleted


if __name__ == "__main__":
    main()
